### PR TITLE
Use Joblib to parallelize optimization

### PR DESCRIPTION
**Pull request by AnesBenmerzoug**
This PR is currently a work in-progress that I wanted to share before proceeding further.
It's better to get feedback at early stage, because maybe the Optuna team does not think that is a good addition to the project.

Using Joblib allows the use of different backends to run the optimization:
 - Threading
 - Multiprocessing
 - Loky
 - Dask

Including any future backends that will be added.

This currently only works with the Threading backend due to the presence of unpicklable Locks in the Study and RDBStorage classes.

The one in the Study class be replaced by a variable that keeps track of the nesting level, which is both cheaper and simpler.

As for the one in the RDBStorage class it is used in the finished trial cache and I haven't figured yet a way to change it.

Feedback would be greatly appreciated.

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@AnesBenmerzoug 

Thank you for your PR.

We also plan to support Multiprocessing, so we'd like to welcome your contribution. `joblib` seems to be a promising approach because it supports various kinds of backends in addition to Multiprocessing.

As you mentioned, we have some issues with design and implementation. Let me help you solve them.

> As for the one in the RDBStorage class it is used in the finished trial cache and I haven't figured yet a way to change it.

I think we can make `RDBStorage` pickable if we remove the unpickable attributes before pickling the storage. It can be implentated using `__setstate__` and `__getstate__` as follows:

```diff
diff --git a/optuna/storages/rdb/storage.py b/optuna/storages/rdb/storage.py
index 8f252a7..194abb0 100644
--- a/optuna/storages/rdb/storage.py
+++ b/optuna/storages/rdb/storage.py
@@ -57,12 +57,14 @@ class RDBStorage(BaseStorage):
 
         self._check_python_version()
 
-        engine_kwargs = engine_kwargs or {}
+        self.engine_kwargs = engine_kwargs or {}
 
-        url = self._fill_storage_url_template(url)
+        self.url = self._fill_storage_url_template(url)
+        self.enable_cache = enable_cache
+        self.skip_compatibility_check = skip_compatibility_check
 
         try:
-            self.engine = create_engine(url, **engine_kwargs)
+            self.engine = create_engine(self.url, **self.engine_kwargs)
         except ImportError as e:
             raise ImportError('Failed to import DB access module for the specified storage URL. '
                               'Please install appropriate one. (The actual import error is: ' +
@@ -73,11 +75,42 @@ class RDBStorage(BaseStorage):
 
         self.logger = optuna.logging.get_logger(__name__)
 
-        self._version_manager = _VersionManager(url, self.engine, self.scoped_session)
+        self._version_manager = _VersionManager(self.url, self.engine, self.scoped_session)
         if not skip_compatibility_check:
             self._version_manager.check_table_schema_compatibility()
 
-        self._finished_trials_cache = _FinishedTrialsCache(enable_cache)
+        self._finished_trials_cache = _FinishedTrialsCache(self.enable_cache)
+
+    def __getstate__(self):
+        # type: () -> Dict[Any, Any]
+
+        state = self.__dict__.copy()
+        del state['scoped_session']
+        del state['engine']
+        del state['logger']
+        del state['_version_manager']
+        del state['_finished_trials_cache']
+        return state
+
+    def __setstate__(self, state):
+        # type: (Dict[Any, Any]) -> None
+
+        self.__dict__.update(state)
+        try:
+            self.engine = create_engine(self.url, **self.engine_kwargs)
+        except ImportError as e:
+            raise ImportError('Failed to import DB access module for the specified storage URL. '
+                              'Please install appropriate one. (The actual import error is: ' +
+                              str(e) + '.)')
+
+        self.scoped_session = orm.scoped_session(orm.sessionmaker(bind=self.engine))
+        models.BaseModel.metadata.create_all(self.engine)
+
+        self.logger = optuna.logging.get_logger(__name__)
+        self._version_manager = _VersionManager(self.url, self.engine, self.scoped_session)
+        if not self.skip_compatibility_check:
+            self._version_manager.check_table_schema_compatibility()
+        self._finished_trials_cache = _FinishedTrialsCache(self.enable_cache)
 
     @staticmethod
     def _check_python_version():
diff --git a/optuna/study.py b/optuna/study.py
index 223f5be..9af3911 100644
--- a/optuna/study.py
+++ b/optuna/study.py
@@ -464,7 +464,7 @@ class Study(BaseStudy):
             if n_jobs is not None and n_jobs > 0:
                 n_jobs = min(n_jobs, n_trials)
 
-        with Parallel(n_jobs=n_jobs, prefer="threads") as parallel:
+        with Parallel(n_jobs=n_jobs, prefer="processes") as parallel:
             # We get n_jobs from joblib to avoid
             # redoing the necessary checks for the different backends
             n_trials_to_enqueue_per_iteration = effective_n_jobs(n_jobs)
```

This trick has been applied to `Study`, so we can run the studies with `RDBStorage` as follows:

```python
>>> import optuna
>>> study = optuna.create_study(storage='sqlite:///example.db')
[I 2019-11-15 11:34:53,748] A new study created with name: no-name-e02e0044-0c43-40ae-ab56-f3cfac9fa244
>>> study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
[I 2019-11-15 11:35:02,179] Finished trial#0 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,181] Finished trial#1 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,492] Finished trial#2 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,494] Finished trial#3 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,675] Finished trial#5 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,678] Finished trial#4 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,850] Finished trial#6 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:02,853] Finished trial#7 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:03,049] Finished trial#8 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
[I 2019-11-15 11:35:03,050] Finished trial#9 resulted in value: 0.7978298005115735. Current best value is 0.7978298005115735 with parameters: {'x': 0.7978298005115735}.
>>> study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
[I 2019-11-15 11:35:15,290] Finished trial#10 resulted in value: 0.053567752867163476. Current best value is 0.053567752867163476 with parameters: {'x': 0.053567752867163476}.
[I 2019-11-15 11:35:15,292] Finished trial#11 resulted in value: 0.053567752867163476. Current best value is 0.053567752867163476 with parameters: {'x': 0.053567752867163476}.
[I 2019-11-15 11:35:15,561] Finished trial#13 resulted in value: 0.0014045606728036383. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:15,562] Finished trial#12 resulted in value: 0.0014045606728036383. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:15,777] Finished trial#15 resulted in value: 0.2567856047684239. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:15,777] Finished trial#14 resulted in value: 0.2567856047684239. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:15,974] Finished trial#17 resulted in value: 0.21394687376471583. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:15,976] Finished trial#16 resulted in value: 0.21394687376471583. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:16,211] Finished trial#18 resulted in value: 0.5144371143314813. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
[I 2019-11-15 11:35:16,219] Finished trial#19 resulted in value: 0.5144371143314813. Current best value is 0.0014045606728036383 with parameters: {'x': 0.0014045606728036383}.
>>> 
```

As you can see, the trials can be executed in parallel, but the results are not as expected.
For instance, the first ten trials returned exactly the same objective values. I think this is because the RandomSampler in each sub-process is always initialized by the same seed. So, I think we need to take care of random states in samplers so as not to sample the same points.

**Comment by AnesBenmerzoug** [comment]
Thanks @toshihikoyanase for the suggestion. I was considering doing that change with the ```___getstate___``` and ```___setstate___``` methods, but I didn't quite know how to do it.

Regarding the issue with parallelism and random state I found this section on joblib's documentation:
https://joblib.readthedocs.io/en/latest/auto_examples/parallel_random_state.html#sphx-glr-auto-examples-parallel-random-state-py
It seems to be taken care of when using the ```threading```, ```loky``` and hopefully the ```dask``` backend ( which a more robust implementation of multiprocessing ), whereas the issue does arise when using the ```multiprocessing``` backend.
I will run your test study once I implement your suggestion for pickling the RBDStorage and post results here.

**Comment by AnesBenmerzoug** [comment]
@toshihikoyanase I ended up using the same trick as for RBDStorage on RandomSampler to make sure each worker process gets a different random number generator when it's unpickled.
I left the default parallel optimization as threading. I think it makes sense as a default and I didn't want to introduce a breaking change.

I ran a few tests to confirm

For threading:

```python
>>> import optuna
>>> study = optuna.create_study(storage='sqlite:///example.db')
[I 2019-11-16 00:13:00,716] A new study created with name: no-name-7ee3045d-29b1-44f6-b960-7d20acda74be
>>> study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
[I 2019-11-16 00:13:11,334] Finished trial#1 resulted in value: 0.718695462200498. Current best value is 0.718695462200498 with parameters: {'x': 0.718695462200498}.
[I 2019-11-16 00:13:11,484] Finished trial#0 resulted in value: 0.7950987477224646. Current best value is 0.718695462200498 with parameters: {'x': 0.718695462200498}.
[I 2019-11-16 00:13:12,146] Finished trial#2 resulted in value: 0.47797773876508065. Current best value is 0.47797773876508065 with parameters: {'x': 0.47797773876508065}.
[I 2019-11-16 00:13:12,303] Finished trial#3 resulted in value: 0.6116867180504431. Current best value is 0.47797773876508065 with parameters: {'x': 0.47797773876508065}.
[I 2019-11-16 00:13:12,847] Finished trial#4 resulted in value: 0.9211815694733351. Current best value is 0.47797773876508065 with parameters: {'x': 0.47797773876508065}.
[I 2019-11-16 00:13:13,100] Finished trial#5 resulted in value: 0.531269291595019. Current best value is 0.47797773876508065 with parameters: {'x': 0.47797773876508065}.
[I 2019-11-16 00:13:13,692] Finished trial#6 resulted in value: 0.3810894177818259. Current best value is 0.3810894177818259 with parameters: {'x': 0.3810894177818259}.
[I 2019-11-16 00:13:13,920] Finished trial#7 resulted in value: 0.9705211783180846. Current best value is 0.3810894177818259 with parameters: {'x': 0.3810894177818259}.
[I 2019-11-16 00:13:14,550] Finished trial#8 resulted in value: 0.402507591463264. Current best value is 0.3810894177818259 with parameters: {'x': 0.3810894177818259}.
[I 2019-11-16 00:13:14,842] Finished trial#9 resulted in value: 0.5764280056821347. Current best value is 0.3810894177818259 with parameters: {'x': 0.3810894177818259}.
```

For Multiprocessing:

```python
>>> import optuna
>>> import joblib
>>> study = optuna.create_study(storage='sqlite:///example.db')
[I 2019-11-16 00:18:02,463] A new study created with name: no-name-d8f7b000-5b85-4d9f-9007-9d24f31aa488
>>> with joblib.parallel_backend("multiprocessing"):
...     study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
...
_pickle.PicklingError: Can't pickle <function <lambda> at 0x7f8be27d5e18>: attribute lookup <lambda> on __main__ failed
```

Unfortunately using multiprocessing fails because with the default pickle module certain functions cannot be pickled.

For Loky:

```python
>>> import optuna
>>> import joblib
>>> study = optuna.create_study(storage='sqlite:///example.db')
[I 2019-11-16 00:19:28,386] A new study created with name: no-name-56935782-6c64-49ee-a558-19c0a8e6463f
>>> with joblib.parallel_backend("loky"):
...     study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
... 
[I 2019-11-16 00:19:39,925] Finished trial#0 resulted in value: 0.6654769953420677. Current best value is 0.6654769953420677 with parameters: {'x': 0.6654769953420677}.
[I 2019-11-16 00:19:40,160] Finished trial#1 resulted in value: 0.7006295138879929. Current best value is 0.6654769953420677 with parameters: {'x': 0.6654769953420677}.
[I 2019-11-16 00:19:40,845] Finished trial#2 resulted in value: 0.1334620732534112. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:41,013] Finished trial#3 resulted in value: 0.5176763955979402. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:41,423] Finished trial#4 resulted in value: 0.7112428831285834. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:41,866] Finished trial#5 resulted in value: 0.8256315270138256. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:42,416] Finished trial#6 resulted in value: 0.8187597464537167. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:42,685] Finished trial#7 resulted in value: 0.757438626415114. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:43,352] Finished trial#8 resulted in value: 0.5478803247091305. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
[I 2019-11-16 00:19:43,475] Finished trial#9 resulted in value: 0.7376578932458349. Current best value is 0.1334620732534112 with parameters: {'x': 0.1334620732534112}.
```

For Dask:

```python
>>> import optuna
>>> import joblib
>>> from distributed import LocalCluster
>>> from distributed import Client
>>> study = optuna.create_study(storage='sqlite:///example.db')
[I 2019-11-16 00:20:58,503] A new study created with name: no-name-bfbbcdcf-63c9-4fee-88fe-648b4dea4039
>>> cluster = LocalCluster(n_workers=2)
>>> client = Client(cluster)
>>> with joblib.parallel_backend("dask"):
...     study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=-1)
... 
```
In this case, n_jobs can be set to -1 or None to use all Dask workers.
Unfortunately, the Dask example does not work with sqlite so I couldn't show numbers for it.

Both Loky and Dask use cloudpickle instead of pickle for serializing objects.

**Comment by toshihikoyanase** [comment]
Thank you for your modification.

> each worker process gets a different random number generator

According to [the document](https://joblib.readthedocs.io/en/latest/auto_examples/parallel_random_state.html#sphx-glr-auto-examples-parallel-random-state-py) you mentioned, it is a typical approach to this issue. I'm still investigating the alternative approaches, but this approach seems to be a good candidate because of the simplicity of implementation.

BTW, I found an issue in the current implementation.
If I set the random seed, all trials suggested the same parameters. I think this is because the current `RandomSampler` always initializes a random number generator using the same seed.

```python
>>> import optuna
>>> import joblib
>>> study = optuna.create_study(storage='sqlite:///example.db', sampler=optuna.samplers.RandomSampler(seed=1))
[I 2019-11-20 15:47:06,929] A new study created with name: no-name-df576cb2-5d7c-491f-a6fc-5f97b06f8f33
>>> with joblib.parallel_backend("loky"):
...     study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
... 
[I 2019-11-20 15:47:17,089] Finished trial#1 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,094] Finished trial#0 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,393] Finished trial#2 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,402] Finished trial#3 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,607] Finished trial#4 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,611] Finished trial#5 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,803] Finished trial#7 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,804] Finished trial#6 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,991] Finished trial#8 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
[I 2019-11-20 15:47:17,991] Finished trial#9 resulted in value: 0.417022004702574. Current best value is 0.417022004702574 with parameters: {'x': 0.417022004702574}.
```

One solution is to ignore random seeds when `RandomSampler` runs in a parallel environment. 
I'm not sure if `RandomSampler` can detect whether it runs in parallel or not. If possible, I think it is reasonable because we can not ensure the reproducibility of studies when we run trials in parallel.
What do you think of it?

**Comment by AnesBenmerzoug** [comment]
@toshihikoyanase I agree that something has to be done for that case. 
It would be useless if a user wants to parallelize the optimization but can't make the process repeatable, if he/she desires.
I would like to avoid simply reseeding ```RandomSampler``` with a random seed when run in a parallel environment because that would break the repeatability.
I found [this comment](https://github.com/numpy/numpy/issues/9650#issuecomment-388045985) on issue [#9650](https://github.com/numpy/numpy/issues/9650) raised in ```numpy```'s repository that describes what I think would be the best approach.
In short, when run a in parallel environment, each ```RandomSampler``` instance gets a new seed sampled from the main thread's random number generator. i.e. when pickling ```RandomSampler``` we sample a new seed from the given user seed.
The downside of this approach is that if we someone pickles a ```RandomSampler``` instance they will get a new seed each time they do so and that, at the very least, should be described in the docs.
What are your thoughts?

**Comment by AnesBenmerzoug** [comment]
After implementing the changes I wrote about, I get the following results when running the test study:

```python
>>> import optuna
>>> import joblib
>>> study = optuna.create_study(storage='sqlite:///example.db', sampler=optuna.samplers.RandomSampler(seed=1))
[I 2019-11-21 19:14:36,413] A new study created with name: no-name-a47b0ebe-0e1a-4e5c-89b0-37906872c442
>>> with joblib.parallel_backend("loky"):
...      study.optimize(lambda t: t.suggest_uniform('x', 0, 1), n_trials=10, n_jobs=2)
... 
[I 2019-11-21 19:14:46,345] Finished trial#0 resulted in value: 0.23955062740611444. Current best value is 0.23955062740611444 with parameters: {'x': 0.23955062740611444}.
[I 2019-11-21 19:14:46,604] Finished trial#1 resulted in value: 0.31449730371913875. Current best value is 0.23955062740611444 with parameters: {'x': 0.23955062740611444}.
[I 2019-11-21 19:14:47,448] Finished trial#3 resulted in value: 0.8841708498826827. Current best value is 0.23955062740611444 with parameters: {'x': 0.23955062740611444}.
[I 2019-11-21 19:14:47,496] Finished trial#2 resulted in value: 0.5399214748497464. Current best value is 0.23955062740611444 with parameters: {'x': 0.23955062740611444}.
[I 2019-11-21 19:14:48,086] Finished trial#5 resulted in value: 0.8776851319206267. Current best value is 0.23955062740611444 with parameters: {'x': 0.23955062740611444}.
[I 2019-11-21 19:14:48,413] Finished trial#4 resulted in value: 0.7034400418549064. Current best value is 0.23955062740611444 with parameters: {'x': 0.23955062740611444}.
[I 2019-11-21 19:14:49,041] Finished trial#7 resulted in value: 0.1465350449525178. Current best value is 0.1465350449525178 with parameters: {'x': 0.1465350449525178}.
[I 2019-11-21 19:14:49,420] Finished trial#6 resulted in value: 0.9675657964283415. Current best value is 0.1465350449525178 with parameters: {'x': 0.1465350449525178}.
[I 2019-11-21 19:14:50,371] Finished trial#9 resulted in value: 0.44742395282759617. Current best value is 0.1465350449525178 with parameters: {'x': 0.1465350449525178}.
[I 2019-11-21 19:14:50,415] Finished trial#8 resulted in value: 0.3161454529854494. Current best value is 0.1465350449525178 with parameters: {'x': 0.1465350449525178}.
```

@toshihikoyanase What do you think of this?

**Comment by sile** [comment]
> It would be useless if a user wants to parallelize the optimization but can't make the process repeatable, if he/she desires.
> I would like to avoid simply reseeding `RandomSampler` with a random seed when run in a parallel environment because that would break the repeatability.
> ...
> In short, when run a in parallel environment, each RandomSampler instance gets a new seed sampled from the main thread's random number generator. i.e. when pickling RandomSampler we sample a new seed from the given user seed.

IMO, the simple reseeding approach is not so bad.
Although the current approach (i.e., "each `RandomSampler` instance gets a new seed sampled from the main thread's random number generator") makes it possible for  `RandomSampler` to reproduce the optimization process, it doesn't apply to other samplers like `TPESampler` and `CmaEsSampler` that consider the previous trial results when sampling new parameters. If you use such a sampler, even though the random number generators have reproducible seeds, the optimization results could change depending on the execution order of the worker threads or processes (the order is inherently uncontrollable).
Because it's impossible to ensure the reproducibility for all samplers, I think that it's reasonable to adopt the reseeding approach for its simplicity.

What do you think of it? > @AnesBenmerzoug @toshihikoyanase

**Comment by toshihikoyanase** [comment]
@sile @ AnesBenmerzoug

>  If you use such a sampler, even though the random number generators have reproducible seeds, the optimization results could change depending on the execution order of the worker threads or processes (the order is inherently uncontrollable).

I made an example to confirm that samplers considering past trials cannot guarantee the reproducibility of experiments. Please take a look at it.

https://colab.research.google.com/drive/1pPxXkFzYSFG4lQE8B6BGTTdAg8_IW2ax

According to the result of `NeighborSampler`, I agree with sile's opinion about the reproducibility.

As for the implementation,  I think that initializing the random number generators without [hierarchical seeding](https://github.com/numpy/numpy/issues/9650#issuecomment-388045985) may be intuitive for users because the RandomSampler is used in the other samplers.

For example, in TPESampler, the RandomSampler is internally used to sample parameter values until the first 10 trials. That means the sampler loses reproducibility in the middle of a study. In that case, users may think the sampler has some bugs.

But I don't think the approach to initialize the RandomSampler is the main topic of this PR, we can keep the current implementation and discuss it after introducing the joblib.

**Comment by AnesBenmerzoug** [comment]
@sile @toshihikoyanase  you both have a point. It seems like it's just a matter of choice.
I will change the ```RandomSampler``` to use a random seed when pickled.
As long as its documented somewhere the users won't be confused.

**Comment by toshihikoyanase** [comment]
@AnesBenmerzoug Thank you for your modification of `RandomSampler`.
I examined the implementation in terms of parallelization efficiency and I found a problem with it.

In this implementation, a master thread/process invokes `n_jobs` trials at a time, and it waits for the slowest trials.

I made a notebook to confirm this phenomenon as follows:
https://colab.research.google.com/drive/10qpvJPyfD8CneV9y3eDjAAc2_R4VyCki

The trials sleep in `0` to `9` secs depending on the corresponding threads. The following log message shows that two threads always start trials at the same time, and that means that thread `140331400050432` (which sleeps in 2 secs) waits for thread `140331400050432` (which sleeps in 8 secs) even if it finished the evaluation.

image

This behavior can be a problem when users parallelize ML model training because execution times are usually different trial by trial, and most of threads/processes will be idle.

So, I created a prototype of an [alternative implementation in my fork repository](https://github.com/toshihikoyanase/optuna/commit/eddd3ee2569c82d6562869f8c147e7338dafbc5c).

The main change related to `joblib.Parallel` is as follows:

```python
                    _iter = range(n_trials) if n_trials is not None else iter(int, 0)
                    parallel(
                        delayed(self._optimize_sequential)
                        (func, 1, timeout, catch, callbacks, gc_after_trial, time_start)
                        for _ in _iter)
```

It passes `range(n_trials)` to `parallel` at a time instead of passing `range(n_jobs)` repeatedly. I think it can start trials without waiting for the other trials.

I also created a notebook for this implementation as follows:
https://colab.research.google.com/drive/1FJ78qtc8es48AQjRFxH6CG162LMY4-7_

The log message shows that the thread `140151838209792` (which sleeps in 2 secs) starts next trials without waiting for the thread `140151829817088` (which sleeps in 8 secs).

image

If this implementation is acceptable, please reflect it to your branch.

**Comment by AnesBenmerzoug** [comment]
@toshihikoyanase thanks for the thorough testing.
I had that in mind when I started this PR, but then I got distracted by the serialization and the seed issue.
I am currently away from my laptop for a few days. So I will make those changes as soon as I come back.

**Comment by AnesBenmerzoug** [comment]
@toshihikoyanase I have fixed the code formatting. Anything else to fix before merging?

File: tests/storages_tests/rdb_tests/test_storage.py [hunk 1]
```diff
@@ -251,6 +251,7 @@ def create_test_storage(enable_cache=True, engine_kwargs=None):
 
 def test_pickle_storage():
     # type: () -> None
+
```
Comment (toshihikoyanase) [inline tests/storages_tests/rdb_tests/test_storage.py]: [A CI job](https://circleci.com/gh/optuna/optuna/23716?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) failed due to the whitespaces in this line. This CI job is mandatory, and we can not leave this issue even if it is trivial. So, please fix it.

File: optuna/study.py [hunk 2]
```diff
@@ -269,19 +268,26 @@ def optimize(
...
             if n_jobs == 1:
                 self._optimize_sequential(func, n_trials, timeout, catch, callbacks,
-                                          gc_after_trial)
+                                          gc_after_trial, None)
             else:
-                self._optimize_parallel(func, n_trials, timeout, n_jobs, catch, callbacks,
-                                        gc_after_trial)
+                time_start = datetime.datetime.now()
+                with Parallel(n_jobs=n_jobs, prefer="threads") as parallel:
+                    _iter = range(n_trials) if n_trials is not None else iter(int, 0)
```
Comment (sile) [inline optuna/study.py]: If `n_trials` is `None`, the study doesn't run any trials because `iter(int, 0)` returns an empty iterator.

How about fixing the code to create `_iter` as follows:
```python
if n_trials is not None:
    _iter = range(n_trials)
elif timeout is not None:
    is_timeout = lambda: (datetime.datetime.now() - time_start).total_seconds() > timeout
    _iter = iter(is_timeout, True)
else:
    # The following expression makes an iterator that never ends.
    _iter = iter(int, 1)
```

File: optuna/study.py [hunk 3]
```diff
@@ -233,7 +231,7 @@ def optimize(
             func,  # type: ObjectiveFuncType
             n_trials=None,  # type: Optional[int]
             timeout=None,  # type: Optional[float]
-            n_jobs=1,  # type: int
+            n_jobs=1,  # type: Optional[int]
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply optuna/study.py]: You're right. It only makes it more confusing to keep than to remove it.

File: optuna/study.py (hunk 2 above)
Comment (AnesBenmerzoug) [reply optuna/study.py]: Nice catch. I did not test that part of the code.
Thanks for the suggested fix. I'll add it right away.
